proxy/grpcproxy/lease.go

```diff
@@ -47,41 +98,252 @@ func (lp *leaseProxy) LeaseTimeToLive(ctx context.Context, rr *pb.LeaseTimeToLiv
+		}
+	}
+	var (
+		recvLoopStopc = make(chan struct{})
+		sendLoopStopc = make(chan struct{})
+		tearDownStopc = make(chan struct{})
+	)
 	go func() {
-		// Cancel the context attached to lc to unblock lc.Recv when
-		// this routine returns on error.
-		defer cancel()
-
-		for {
-			// stream.Recv will be unblock when the loop in the parent routine
-			// returns on error.
-			rr, err := stream.Recv()
-			if err != nil {
-				return
-			}
-			err = lc.Send(rr)
-			if err != nil {
-				return
-			}
+		defer func() { close(recvLoopStopc) }()
+		lps.toErrCh(lps.recvLoop())
+	}()
+
+	go func() {
+		defer func() { close(sendLoopStopc) }()
+		lps.toErrCh(lps.sendLoop())
+	}()
+
+	// tears down LeaseKeepAlive stream if leader goes down or entire leaseProxy is terminated.
+	go func() {
+		defer func() { close(tearDownStopc) }()
+		select {
+		case <-leaderc:
+		case <-ctx.Done():
+		case <-lp.ctx.Done():
 		}
 	}()
 
-	for {
-		rr, err := lc.Recv()
+	var err error
+	select {
+	case <-recvLoopStopc:
+	case <-tearDownStopc:
+	case err = <-lps.errCh:
+	}
+	cancel()
+
+	// recv/send may only shutdown after function exits;
+	// this goroutine notifies lease proxy that the stream is through
+	go func() {
+		<-recvLoopStopc
+		//
+		lps.close()
+		<-tearDownStopc
+		<-sendLoopStopc
+		lp.wg.Done()
+	}()
+
+	select {
+	case <-leaderc:
+		return rpctypes.ErrNoLeader
+	default:
 		if err != nil {
 			return err
 		}
-		err = stream.Send(rr)
+		return ctx.Err()
+	}
+}
+
+type leaseProxyStream struct {
+	stream pb.Lease_LeaseKeepAliveServer
+
+	leaseCli clientv3.Lease
+
+	client *clientv3.Client
+
+	wg sync.WaitGroup
+	// mu protects keepAliveLeases
+	mu sync.RWMutex
+	// keepAliveLeases map keeps track of on-going lease renewal IDs their atomic counters.
+	// the counter is used to control a 1:1 keepalive req and resp ratio.
+	// e.g counter++ when recv a KeepAlive req and counter-- when sending a KeepAlive resp to client
+	// when counter == 0, don't send KeepAlive resp to the client
+	keepAliveLeases map[int64]*atomicCounter
+	// respc is a channel to receive lease keepalive resp from etcd backend
+	respc chan *pb.LeaseKeepAliveResponse
+
+	ctx context.Context
+
+	cancel func()
+	// errCh must receive a non-nil error because that error
+	// will be returned to the calling client to indicate a failure
+	// happened at the backend
+	errCh chan error
+}
+
+func (lps *leaseProxyStream) recvLoop() error {
+	for lps.ctx.Err() == nil {
+		rr, err := lps.stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
 		if err != nil {
 			return err
 		}
+		lps.serveKeepAliveRequest(rr)
+	}
+	return lps.ctx.Err()
+}
+
+func (lps *leaseProxyStream) serveKeepAliveRequest(rr *pb.LeaseKeepAliveRequest) {
+	lps.mu.Lock()
+	defer lps.mu.Unlock()
+	canSendRespC, ok := lps.keepAliveLeases[rr.ID]
+	if !ok {
+		canSendRespC = &atomicCounter{}
+		lps.keepAliveLeases[rr.ID] = canSendRespC
+		lps.wg.Add(1)
+		go func() {
+			defer lps.wg.Done()
+			lps.toErrCh(lps.keepAliveLoop(rr.ID))
```

Comment:
this should probably just call `lps.cancel()` instead of piping an error through